src/ace/managers/advancedsprite.c

```diff
@@ -0,0 +1,259 @@
+             //attached sprite
+             pAdvancedSprite->pSprites[i] = spriteAdd(ubChannelIndex+i, pAdvancedSprite->pAnimBitmap[i]);
+             spriteSetAttached(pAdvancedSprite->pSprites[i],1);
+         } else {
+             pAdvancedSprite->pSprites[i] = spriteAdd(ubChannelIndex+i, pAdvancedSprite->pAnimBitmap[i]);
+         }      
+     }
+ 
+     return pAdvancedSprite;
+ }
+ 
+ void advancedSpriteRemove(tAdvancedSprite *pAdvancedSprite) {
+     systemUse();
+     for (UWORD i = 0; i < pAdvancedSprite->ubSpriteCount; i++) {
+         spriteRemove(pAdvancedSprite->pSprites[i]);
+     }
+     for (UWORD i = 0; i < pAdvancedSprite->uwAnimCount; i++) {
+         bitmapDestroy(pAdvancedSprite->pAnimBitmap[i]);
+     }
+     memFree(pAdvancedSprite, sizeof(*pAdvancedSprite));
+     systemUnuse();
+ }
+ 
+ void advancedSpriteSetEnabled(tAdvancedSprite *pAdvancedSprite, UBYTE isEnabled) {
+     for (UWORD i = 0; i < pAdvancedSprite->ubSpriteCount; i++) {
+         spriteSetEnabled(pAdvancedSprite->pSprites[i], isEnabled);
+     }
+ }
+ 
+ void advancedSpriteSetPos(tAdvancedSprite *pAdvancedSprite,WORD wX, WORD wY) {
+     pAdvancedSprite->wX = wX;
+     pAdvancedSprite->wY = wY;
+     pAdvancedSprite->isHeaderToBeUpdated = 1;
+ }
+ 
+ void advancedSpriteSetPosX(tAdvancedSprite *pAdvancedSprite,WORD wX) {
+     pAdvancedSprite->wX = wX;
+     pAdvancedSprite->isHeaderToBeUpdated = 1;
+ }
+ 
+ void advancedSpriteSetPosY(tAdvancedSprite *pAdvancedSprite, WORD wY) {
+     pAdvancedSprite->wY = wY;
+     pAdvancedSprite->isHeaderToBeUpdated = 1;
+ }
+ 
+ void advancedSpriteSetFrame(tAdvancedSprite *pAdvancedSprite, UWORD animFrame) {
+     if (animFrame >= pAdvancedSprite->uwAnimCount) {
+         logWrite("ERR: Invalid animation index %hu\n", animFrame);
+         return;
+     }
+     pAdvancedSprite->uwAnimFrame=animFrame;
+     UWORD animIndex = animFrame << ((pAdvancedSprite->ubByteWidth == 4) + pAdvancedSprite->is4PP);
+     for (UWORD i = 0; i < pAdvancedSprite->ubSpriteCount; i++) {
+         spriteSetBitmap(pAdvancedSprite->pSprites[i], pAdvancedSprite->pAnimBitmap[animIndex+i]);
+         pAdvancedSprite->isHeaderToBeUpdated = 1; // To force header rewrite
+     }    
+ }
+ 
+ void advancedSpriteProcessChannel(tAdvancedSprite *pAdvancedSprite) {
+     for (UWORD i = 0; i < pAdvancedSprite->ubSpriteCount; i++) {
+         spriteProcessChannel(pAdvancedSprite->ubChannelIndex + i);
+     }
+ }
+ 
+ void advancedSpriteProcess(tAdvancedSprite *pAdvancedSprite) {
+     if(!pAdvancedSprite->isHeaderToBeUpdated) {
+         return;
+     }
+     for (UWORD i = 0; i < pAdvancedSprite->ubSpriteCount; i++) {
+         if (( pAdvancedSprite->ubByteWidth==4) && (( (i>1) && (pAdvancedSprite->is4PP== 1) ) || ( (i==1) && (pAdvancedSprite->is4PP==0) )))
+         {
+             pAdvancedSprite->pSprites[i]->wX = pAdvancedSprite->wX+SPRITE_WIDTH;;
+         } else {
+             pAdvancedSprite->pSprites[i]->wX = pAdvancedSprite->wX;
+         }
```

Comment:
I think it'd be better to have something like pAdvancedSprite->pAddX[i] being set to either 0 or SPRITE_WIDTH and just have
`pAdvancedSprite->pSprites[i]->wX = pAdvancedSprite->wX + pAdvancedSprite->pAddX[i];`. Saves us that ugly `if`.

Can be done at a later time, let me know how you see it.

Reply:
By ugly I meant performance-wise ;)

Note that there will be 4-5 checks in resulting code, and for the sprite consisting of 4 hardware sprites that would mean 16-20 checks instead of 4x loads of x offset from memory.

But that can be optimized later on.